File: bot/customer-bot/src/solana.ts
/**
 * Solana RPC helpers for the customer-facing bot.
 *
 * All functions are read-only — they fetch on-chain account data
 * but never submit transactions. No keypair required.
 *
 * Uses MANUAL binary deserialization (no @coral-xyz/anchor dependency)
 * so it works reliably in Cloudflare Workers runtime.
 */

import { Connection, PublicKey } from "@solana/web3.js";
import type { BotConfig } from "./config";
import {
  deriveMainPDAs,
  deriveQPPDAs,
  deriveDrawResultPDA,
  deriveQPDrawResultPDA,
  deriveUserStatsPDA,
} from "./config";

// ============================================================================
// MINIMAL BORSH DESERIALIZER
// ============================================================================
// Cloudflare Workers don't reliably support @coral-xyz/anchor's BorshCoder
// due to Node.js Buffer compatibility issues. Instead we manually parse
// the binary layouts matching the exact Rust struct definitions.

class BorshReader {
  private buf: Uint8Array;
  private offset: number;

  constructor(buf: Uint8Array, offset: number = 0) {
    this.buf = buf;
    this.offset = offset;
  }

  readU8(): number {
    return this.buf[this.offset++];
  }

  readU16(): number {
    const lo = this.buf[this.offset++];
    const hi = this.buf[this.offset++];
    return lo | (hi << 8);
  }

  readU32(): number {
    const b0 = this.buf[this.offset++];
    const b1 = this.buf[this.offset++];
    const b2 = this.buf[this.offset++];
    const b3 = this.buf[this.offset++];
    return (b0 | (b1 << 8) | (b2 << 16) | (b3 << 24)) >>> 0;
  }

  readU64(): bigint {
    let lo = 0n;
    let hi = 0n;
    for (let i = 0; i < 4; i++) {
      lo |= BigInt(this.buf[this.offset++]) << BigInt(i * 8);
    }
    for (let i = 0; i < 4; i++) {
      hi |= BigInt(this.buf[this.offset++]) << BigInt(i * 8);
    }
    return lo | (hi << 32n);
  }

  readI64(): bigint {
    const unsigned = this.readU64();
    // Sign-extend
    if (unsigned & (1n << 63n)) {
      return unsigned - (1n << 64n);
    }
    return unsigned;
  }

  readBool(): boolean {
    return this.buf[this.offset++] !== 0;
  }

  readPubkey(): PublicKey {
    const bytes = this.buf.slice(this.offset, this.offset + 32);
    this.offset += 32;
    return new PublicKey(bytes);
  }

  readOptionPubkey(): PublicKey | null {
    const discriminant = this.readU8();
    if (discriminant === 0) return null;
    return this.readPubkey();
  }

  readFixedBytes(len: number): Uint8Array {
    const bytes = this.buf.slice(this.offset, this.offset + len);
    this.offset += len;
    return bytes;
  }

  readU8Array(len: number): number[] {
    const arr: number[] = [];
    for (let i = 0; i < len; i++) {
      arr.push(this.readU8());
    }
    return arr;
  }

  skip(bytes: number): void {
    this.offset += bytes;
  }
}

// ============================================================================
// ACCOUNT DESERIALIZERS — exact field-order matches Rust #[account] structs
// ============================================================================

const ANCHOR_DISCRIMINATOR_LEN = 8;

// --- LotteryState (mazelprotocol) ---
// See: programs/mazelprotocol/src/state.rs
function decodeLotteryState(data: Uint8Array): MainLotteryState {
  const r = new BorshReader(data, ANCHOR_DISCRIMINATOR_LEN);

  return {
    authority: r.readPubkey().toBase58(),
    pendingAuthority: r.readOptionPubkey()?.toBase58() ?? null,
    switchboardQueue: r.readPubkey().toBase58(),
    currentRandomnessAccount: r.readPubkey().toBase58(),
    currentDrawId: r.readU64(),
    jackpotBalance: r.readU64(),
    reserveBalance: r.readU64(),
    insuranceBalance: r.readU64(),
    fixedPrizeBalance: r.readU64(),
    ticketPrice: r.readU64(),
    houseFeeBps: r.readU16(),
    jackpotCap: r.readU64(),
    seedAmount: r.readU64(),
    softCap: r.readU64(),
    hardCap: r.readU64(),
    nextDrawTimestamp: r.readI64(),
    drawInterval: r.readI64(),
    commitSlot: r.readU64(),
    commitTimestamp: r.readI64(),
    currentDrawTickets: r.readU64(),
    totalTicketsSold: r.readU64(),
    totalPrizesPaid: r.readU64(),
    totalPrizesCommitted: r.readU64(),
    isDrawInProgress: r.readBool(),
    isAwaitingFinalization: r.readBool(),
    isRolldownActive: r.readBool(),
    isPaused: r.readBool(),
    isFunded: r.readBool(),
    version: r.readU8(),
    bump: r.readU8(),
    configTimelockEnd: r.readI64(),
    pendingConfigHash: r.readFixedBytes(32),
    emergencyTransferTotal: r.readU64(),
    emergencyTransferWindowStart: r.readI64(),
    maxRolldownTickets: r.readU64(),
  };
}

// --- DrawResult (mazelprotocol) ---
// See: programs/mazelprotocol/src/state.rs
function decodeDrawResult(data: Uint8Array): MainDrawResult {
  const r = new BorshReader(data, ANCHOR_DISCRIMINATOR_LEN);

  return {
    drawId: r.readU64(),
    winningNumbers: r.readU8Array(6),
    randomnessProof: r.readFixedBytes(32),
    timestamp: r.readI64(),
    totalTickets: r.readU64(),
    wasRolldown: r.readBool(),
    match6Winners: r.readU32(),
    match5Winners: r.readU32(),
    match4Winners: r.readU32(),
    match3Winners: r.readU32(),
    match2Winners: r.readU32(),
    match6PrizePerWinner: r.readU64(),
    match5PrizePerWinner: r.readU64(),
    match4PrizePerWinner: r.readU64(),
    match3PrizePerWinner: r.readU64(),
    match2PrizePerWinner: r.readU64(),
    isExplicitlyFinalized: r.readBool(),
    totalCommitted: r.readU64(),
    totalReclaimed: r.readU64(),
    bump: r.readU8(),
  };
}

// --- UserStats (mazelprotocol) ---
// See: programs/mazelprotocol/src/state.rs
function decodeUserStats(data: Uint8Array): UserStats {
  const r = new BorshReader(data, ANCHOR_DISCRIMINATOR_LEN);

  return {
    wallet: r.readPubkey().toBase58(),
    totalTickets: r.readU64(),
    totalSpent: r.readU64(),
    totalWon: r.readU64(),
    currentStreak: r.readU32(),
    bestStreak: r.readU32(),
    jackpotWins: r.readU32(),
    lastDrawParticipated: r.readU64(),
    ticketsThisDraw: r.readU64(),
    freeTicketsAvailable: r.readU32(),
    bump: r.readU8(),
  };
}

// --- QuickPickState (quickpick) ---
// See: programs/quickpick/src/state.rs
function decodeQuickPickState(data: Uint8Array): QPLotteryState {
  const r = new BorshReader(data, ANCHOR_DISCRIMINATOR_LEN);

  const state: QPLotteryState = {
    currentDraw: r.readU64(),
    ticketPrice: r.readU64(),
    pickCount: r.readU8(),
    numberRange: r.readU8(),
    houseFeeBps: r.readU16(),
    drawInterval: r.readI64(),
    nextDrawTimestamp: r.readI64(),
    jackpotBalance: r.readU64(),
    softCap: r.readU64(),
    hardCap: r.readU64(),
    seedAmount: r.readU64(),
    match4Prize: r.readU64(),
    match3Prize: r.readU64(),
    currentDrawTickets: r.readU64(),
    prizePoolBalance: r.readU64(),
    insuranceBalance: r.readU64(),
    reserveBalance: r.readU64(),
    totalTicketsSold: r.readU64(),
    totalPrizesPaid: r.readU64(),
    currentRandomnessAccount: r.readPubkey().toBase58(),
    commitSlot: r.readU64(),
    commitTimestamp: r.readI64(),
    isDrawInProgress: r.readBool(),
    isAwaitingFinalization: r.readBool(),
    isRolldownPending: r.readBool(),
    isPaused: r.readBool(),
    isFunded: r.readBool(),
    bump: r.readU8(),
    configTimelockEnd: r.readI64(),
    pendingConfigHash: r.readFixedBytes(32),
    emergencyTransferTotal: r.readU64(),
    emergencyTransferWindowStart: r.readI64(),
  };

  return state;
}

// --- QuickPickDrawResult (quickpick) ---
// See: programs/quickpick/src/state.rs
function decodeQPDrawResult(data: Uint8Array): QPDrawResult {
  const r = new BorshReader(data, ANCHOR_DISCRIMINATOR_LEN);

  return {
    drawId: r.readU64(),
    winningNumbers: r.readU8Array(5),
    randomnessProof: r.readFixedBytes(32),
    timestamp: r.readI64(),
    totalTickets: r.readU64(),
    wasRolldown: r.readBool(),
    match5Winners: r.readU32(),
    match4Winners: r.readU32(),
    match3Winners: r.readU32(),
    match5PrizePerWinner: r.readU64(),
    match4PrizePerWinner: r.readU64(),
    match3PrizePerWinner: r.readU64(),
    isExplicitlyFinalized: r.readBool(),
    bump: r.readU8(),
  };
}

// ============================================================================
// TYPES
// ============================================================================

export interface MainLotteryState {
  authority: string;
  pendingAuthority: string | null;
  switchboardQueue: string;
  currentRandomnessAccount: string;
  currentDrawId: bigint;
  jackpotBalance: bigint;
  reserveBalance: bigint;
  insuranceBalance: bigint;
  fixedPrizeBalance: bigint;
  ticketPrice: bigint;
  houseFeeBps: number;
  jackpotCap: bigint;
  seedAmount: bigint;
  softCap: bigint;
  hardCap: bigint;
  nextDrawTimestamp: bigint;
  drawInterval: bigint;
  commitSlot: bigint;
  commitTimestamp: bigint;
  currentDrawTickets: bigint;
  totalTicketsSold: bigint;
  totalPrizesPaid: bigint;
  totalPrizesCommitted: bigint;
  isDrawInProgress: boolean;
  isAwaitingFinalization: boolean;
  isRolldownActive: boolean;
  isPaused: boolean;
  isFunded: boolean;
  version: number;
  bump: number;
  configTimelockEnd: bigint;
  pendingConfigHash: Uint8Array;
  emergencyTransferTotal: bigint;
  emergencyTransferWindowStart: bigint;
  maxRolldownTickets: bigint;
}

export interface MainDrawResult {
  drawId: bigint;
  winningNumbers: number[];
  randomnessProof: Uint8Array;
  timestamp: bigint;
  totalTickets: bigint;
  wasRolldown: boolean;
  match6Winners: number;
  match5Winners: number;
  match4Winners: number;
  match3Winners: number;
  match2Winners: number;
  match6PrizePerWinner: bigint;
  match5PrizePerWinner: bigint;
  match4PrizePerWinner: bigint;
  match3PrizePerWinner: bigint;
  match2PrizePerWinner: bigint;
  isExplicitlyFinalized: boolean;
  totalCommitted: bigint;
  totalReclaimed: bigint;
  bump: number;
}

export interface UserStats {
  wallet: string;
  totalTickets: bigint;
  totalSpent: bigint;
  totalWon: bigint;
  currentStreak: number;
  bestStreak: number;
  jackpotWins: number;
  lastDrawParticipated: bigint;
  ticketsThisDraw: bigint;
  freeTicketsAvailable: number;
  bump: number;
}

export interface QPLotteryState {
  currentDraw: bigint;
  ticketPrice: bigint;
  pickCount: number;
  numberRange: number;
  houseFeeBps: number;
  drawInterval: bigint;
  nextDrawTimestamp: bigint;
  jackpotBalance: bigint;
  softCap: bigint;
  hardCap: bigint;
  seedAmount: bigint;
  match4Prize: bigint;
  match3Prize: bigint;
  currentDrawTickets: bigint;
  prizePoolBalance: bigint;
  insuranceBalance: bigint;
  reserveBalance: bigint;
  totalTicketsSold: bigint;
  totalPrizesPaid: bigint;
  currentRandomnessAccount: string;
  commitSlot: bigint;
  commitTimestamp: bigint;
  isDrawInProgress: boolean;
  isAwaitingFinalization: boolean;
  isRolldownPending: boolean;
  isPaused: boolean;
  isFunded: boolean;
  bump: number;
  configTimelockEnd: bigint;
  pendingConfigHash: Uint8Array;
  emergencyTransferTotal: bigint;
  emergencyTransferWindowStart: bigint;
}

export interface QPDrawResult {
  drawId: bigint;
  winningNumbers: number[];
  randomnessProof: Uint8Array;
  timestamp: bigint;
  totalTickets: bigint;
  wasRolldown: boolean;
  match5Winners: number;
  match4Winners: number;
  match3Winners: number;
  match5PrizePerWinner: bigint;
  match4PrizePerWinner: bigint;
  match3PrizePerWinner: bigint;
  isExplicitlyFinalized: boolean;
  bump: number;
}

// ============================================================================
// CONNECTION MANAGEMENT
// ============================================================================

let _connection: Connection | undefined;
let _config: BotConfig | undefined;

export function initSolana(config: BotConfig): void {
  _config = config;
  _connection = new Connection(config.rpcUrl, {
    commitment: config.commitment,
    confirmTransactionInitialTimeout: 30_000,
  });
}

function getConnection(): Connection {
  if (!_connection) throw new Error("Solana not initialized. Call initSolana().");
  return _connection;
}

function getConfig(): BotConfig {
  if (!_config) throw new Error("Solana not initialized. Call initSolana().");
  return _config;
}

// ============================================================================
// PUBLIC FETCHERS
// ============================================================================

export async function fetchMainLotteryState(): Promise<MainLotteryState> {
  const config = getConfig();
  const connection = getConnection();
  const { lotteryState } = deriveMainPDAs(config.mainProgramId);
  const accountInfo = await connection.getAccountInfo(lotteryState, config.commitment);

  if (!accountInfo) {
    throw new Error("Main lottery state account not found. Program may not be initialized.");
  }

  return decodeLotteryState(new Uint8Array(accountInfo.data));
}

export async function fetchQPLotteryState(): Promise<QPLotteryState> {
  const config = getConfig();
  const connection = getConnection();
  const { quickPickState } = deriveQPPDAs(config.qpProgramId);
  const accountInfo = await connection.getAccountInfo(quickPickState, config.commitment);

  if (!accountInfo) {
    throw new Error("Quick Pick state account not found. Program may not be initialized.");
  }

  return decodeQuickPickState(new Uint8Array(accountInfo.data));
}

export async function fetchMainDrawResult(
  drawId: number | bigint,
): Promise<MainDrawResult | null> {
  const config = getConfig();
  const connection = getConnection();
  const [drawResultPDA] = deriveDrawResultPDA(drawId, config.mainProgramId);
  const accountInfo = await connection.getAccountInfo(drawResultPDA, config.commitment);

  if (!accountInfo) return null;

  return decodeDrawResult(new Uint8Array(accountInfo.data));
}

export async function fetchQPDrawResult(
  drawId: number | bigint,
): Promise<QPDrawResult | null> {
  const config = getConfig();
  const connection = getConnection();
  const [drawResultPDA] = deriveQPDrawResultPDA(drawId, config.qpProgramId);
  const accountInfo = await connection.getAccountInfo(drawResultPDA, config.commitment);

  if (!accountInfo) return null;

  return decodeQPDrawResult(new Uint8Array(accountInfo.data));
}

export async function fetchUserStats(
  walletAddress: string,
): Promise<UserStats | null> {
  const config = getConfig();
  const connection = getConnection();
  const wallet = new PublicKey(walletAddress);
  const [userStatsPDA] = deriveUserStatsPDA(wallet, config.mainProgramId);
  const accountInfo = await connection.getAccountInfo(userStatsPDA, config.commitment);

  if (!accountInfo) return null;

  return decodeUserStats(new Uint8Array(accountInfo.data));
}

File: customer-bot-rust/src/solana.rs
//! Solana RPC queries — read-only lottery state fetchers.

use crate::config::BotConfig;
use crate::error::Result;
use anchor_lang::AnchorDeserialize;
use solana_client::client_error::{ClientError, ClientErrorKind};
use solana_client::rpc_client::RpcClient;
use solana_client::rpc_request::RpcError;
use solana_pubkey::Pubkey;

pub const LOTTERY_SEED: &[u8] = b"lottery";
pub const DRAW_SEED: &[u8] = b"draw";
pub const QUICK_PICK_SEED: &[u8] = b"quick_pick";
pub const QUICK_PICK_DRAW_SEED: &[u8] = b"quick_pick_draw";

// ---------------------------------------------------------------------------
// IMPORTANT — account layout mirrors
//
// These structs MUST mirror the on-chain Anchor accounts field-for-field
// (borsh is sequential, no alignment padding). The first field of every
// struct is the 8-byte Anchor account discriminator
// (sha256("account:<Name>")[..8]) which is present in the stored data but
// not part of the logical struct.
//
// Reference: programs/mazelprotocol/src/state/lottery_state.rs and
// programs/quickpick/src/state.rs. If the programs ever change these
// layouts, update the mirrors here — the program crates are the single
// source of truth.
// ---------------------------------------------------------------------------

#[derive(AnchorDeserialize, Debug, Clone)]
pub struct LotteryState {
    // Anchor account discriminator (8 bytes)
    _discriminator: [u8; 8],
    _authority: Pubkey,
    // pending_authority: Option<Pubkey> = 1-byte tag + 32-byte pubkey
    _pending_authority: [u8; 33],
    _switchboard_queue: Pubkey,
    _current_randomness_account: Pubkey,
    pub current_draw_id: u64,
    pub jackpot_balance: u64,
    _reserve_balance: u64,
    _insurance_balance: u64,
    _fixed_prize_balance: u64,
    _ticket_price: u64,
    _house_fee_bps: u16,
    _jackpot_cap: u64,
    _seed_amount: u64,
    _soft_cap: u64,
    _hard_cap: u64,
    pub next_draw_timestamp: i64,
    _draw_interval: i64,
    _commit_slot: u64,
    _commit_timestamp: i64,
    pub current_draw_tickets: u64,
    _total_tickets_sold: u64,
    _total_prizes_paid: u64,
    _total_prizes_committed: u64,
    pub is_draw_in_progress: bool,
    _is_awaiting_finalization: bool,
    _is_rolldown_active: bool,
    pub is_paused: bool,
    _is_funded: bool,
    _version: u8,
    _bump: u8,
    _config_timelock_end: i64,
    _pending_config_hash: [u8; 32],
    _emergency_transfer_total: u64,
    _emergency_transfer_window_start: i64,
    _max_rolldown_tickets: u64,
    _sale_target_tickets: u64,
}

#[derive(AnchorDeserialize, Debug, Clone)]
pub struct DrawResult {
    // Anchor account discriminator (8 bytes)
    _discriminator: [u8; 8],
    pub draw_id: u64,
    pub winning_numbers: [u8; 6],
    _randomness_proof: [u8; 32],
    _timestamp: i64,
    pub total_tickets: u64,
    pub was_rolldown: bool,
    pub match6_win: u32,
    pub match5_win: u32,
    pub match4_win: u32,
    pub match3_win: u32,
    pub match2_win: u32,
    pub m6_prize: u64,
    pub m5_prize: u64,
    pub m4_prize: u64,
    pub m3_prize: u64,
    pub m2_prize: u64,
    _is_explicitly_finalized: bool,
    _total_committed: u64,
    _total_reclaimed: u64,
    _bump: u8,
}

#[derive(AnchorDeserialize, Debug, Clone)]
pub struct QpState {
    // Anchor account discriminator (8 bytes)
    _discriminator: [u8; 8],
    pub current_draw: u64,
    _ticket_price: u64,
    _pick_count: u8,
    _number_range: u8,
    _house_fee_bps: u16,
    _draw_interval: i64,
    pub next_draw_timestamp: i64,
    pub jackpot_balance: u64,
    _soft_cap: u64,
    _hard_cap: u64,
    _seed_amount: u64,
    _match_4_prize: u64,
    _match_3_prize: u64,
    pub current_draw_tickets: u64,
    _prize_pool_balance: u64,
    _insurance_balance: u64,
    _reserve_balance: u64,
    _total_tickets_sold: u64,
    _total_prizes_paid: u64,
    _current_randomness_account: Pubkey,
    _commit_slot: u64,
    _commit_timestamp: i64,
    pub is_draw_in_progress: bool,
    _is_awaiting_finalization: bool,
    _is_rolldown_pending: bool,
    pub is_paused: bool,
    _is_funded: bool,
    _bump: u8,
    _config_timelock_end: i64,
    _pending_config_hash: [u8; 32],
    _emergency_transfer_total: u64,
    _emergency_transfer_window_start: i64,
    _sale_target_tickets: u64,
}

#[derive(AnchorDeserialize, Debug, Clone)]
pub struct QpDrawResult {
    // Anchor account discriminator (8 bytes)
    _discriminator: [u8; 8],
    pub draw_id: u64,
    pub winning_numbers: [u8; 5],
    _randomness_proof: [u8; 32],
    _timestamp: i64,
    pub total_tickets: u64,
    pub was_rolldown: bool,
    pub match5_win: u32,
    pub match4_win: u32,
    pub match3_win: u32,
    pub m5_prize: u64,
    pub m4_prize: u64,
    pub m3_prize: u64,
    _is_explicitly_finalized: bool,
    _bump: u8,
}

pub struct Solana {
    rpc: RpcClient,
    cfg: BotConfig,
}

impl Solana {
    pub fn new(cfg: BotConfig) -> Self {
        let rpc = RpcClient::new_with_commitment(cfg.rpc_url.clone(), cfg.commitment);
        Solana { rpc, cfg }
    }

    /// Fetch an account, verify it is owned by the expected program, and
    /// deserialize it as `T`. Failing loudly on owner mismatch prevents
    /// silently parsing garbage when a PDA resolves to the wrong program.
    fn fetch_owned<T: AnchorDeserialize>(
        &self,
        pda: &Pubkey,
        expected_owner: &Pubkey,
        what: &str,
    ) -> Result<T> {
        let acc = self.rpc.get_account(pda)?;
        if acc.owner != *expected_owner {
            return Err(crate::error::Error::Anyhow(anyhow::anyhow!(
                "Account {pda} ({what}) owned by {}, expected {expected_owner}",
                acc.owner
            )));
        }
        let mut data: &[u8] = &acc.data;
        Ok(T::deserialize(&mut data)?)
    }

    pub fn fetch_main_state(&self) -> Result<LotteryState> {
        let (pda, _) = Pubkey::find_program_address(&[LOTTERY_SEED], &self.cfg.main_program_id);
        self.fetch_owned::<LotteryState>(&pda, &self.cfg.main_program_id, "lottery_state")
    }

    pub fn fetch_qp_state(&self) -> Result<QpState> {
        let (pda, _) = Pubkey::find_program_address(&[QUICK_PICK_SEED], &self.cfg.qp_program_id);
        self.fetch_owned::<QpState>(&pda, &self.cfg.qp_program_id, "quick_pick_state")
    }

    pub fn fetch_main_draw(&self, draw_id: u64) -> Result<Option<DrawResult>> {
        let (pda, _) = Pubkey::find_program_address(
            &[DRAW_SEED, &draw_id.to_le_bytes()],
            &self.cfg.main_program_id,
        );
        match self.rpc.get_account(&pda) {
            Ok(_) => Ok(Some(self.fetch_owned::<DrawResult>(
                &pda,
                &self.cfg.main_program_id,
                "main draw_result",
            )?)),
            // Only a genuinely missing account means "draw not found".
            Err(e) if is_account_not_found(&e) => Ok(None),
            // Everything else (RPC outage, rate limit, network failure) is a
            // service problem and must surface as an error, not "not found".
            Err(e) => Err(e.into()),
        }
    }

    pub fn fetch_qp_draw(&self, draw_id: u64) -> Result<Option<QpDrawResult>> {
        let (pda, _) = Pubkey::find_program_address(
            &[QUICK_PICK_DRAW_SEED, &draw_id.to_le_bytes()],
            &self.cfg.qp_program_id,
        );
        match self.rpc.get_account(&pda) {
            Ok(_) => Ok(Some(self.fetch_owned::<QpDrawResult>(
                &pda,
                &self.cfg.qp_program_id,
                "qp draw_result",
            )?)),
            // Only a genuinely missing account means "draw not found".
            Err(e) if is_account_not_found(&e) => Ok(None),
            // Everything else (RPC outage, rate limit, network failure) is a
            // service problem and must surface as an error, not "not found".
            Err(e) => Err(e.into()),
        }
    }
}

/// Returns true when an RPC error means "the account does not exist on
/// chain" (i.e. the draw has not happened yet). Every other error — RPC
/// outage, rate limit, network failure — is a service problem and must be
/// surfaced to the user instead of being masked as "draw not found".
fn is_account_not_found(err: &ClientError) -> bool {
    matches!(
        &err.kind,
        ClientErrorKind::RpcError(RpcError::ForUser(msg)) if msg.starts_with("AccountNotFound")
    ) || matches!(
        &err.kind,
        ClientErrorKind::RpcError(RpcError::RpcResponseError { message, .. })
            if message == "AccountNotFound" || message == "account not found"
    )
}

/// Format USDC lamports as a human-readable string.
pub fn format_usdc(lamports: u64) -> String {
    let usd = lamports as f64 / 1_000_000.0;
    if usd >= 1_000_000.0 {
        format!("${:.2}M", usd / 1_000_000.0)
    } else if usd >= 1_000.0 {
        format!("${:.1}K", usd / 1_000.0)
    } else if usd >= 1.0 {
        format!("${:.2}", usd)
    } else {
        format!("${:.4}", usd)
    }
}

pub fn format_countdown(target: i64) -> String {
    let now = chrono::Utc::now().timestamp();
    let diff = target - now;
    if diff <= 0 {
        return "⏰ Overdue".into();
    }
    let h = diff / 3600;
    let m = (diff % 3600) / 60;
    let s = diff % 60;
    if h > 24 {
        format!("{}d {}h {}m", h / 24, h % 24, m)
    } else if h > 0 {
        format!("{}h {}m {}s", h, m, s)
    } else if m > 0 {
        format!("{}m {}s", m, s)
    } else {
        format!("{}s", s)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn client_error(kind: ClientErrorKind) -> ClientError {
        ClientError { request: None, kind }
    }

    #[test]
    fn missing_account_is_not_found() {
        let err = client_error(ClientErrorKind::RpcError(RpcError::ForUser(
            "AccountNotFound: pubkey=abc".into(),
        )));
        assert!(is_account_not_found(&err));
    }

    #[test]
    fn rpc_outage_is_not_not_found() {
        let outage =
            client_error(ClientErrorKind::RpcError(RpcError::ForUser("node is unhealthy".into())));
        assert!(!is_account_not_found(&outage));
        let custom = client_error(ClientErrorKind::Custom("timeout".into()));
        assert!(!is_account_not_found(&custom));
    }
}
